src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2907,75 +2912,57 @@ void visitArgumentEvaluateAndUnsplit(int argumentIndex, bool assertsTrue, bool a
                             extensionMethodThisArgument: extensionMethodThisArgument);
                     }
                     break;
+                case RefKind.Ref:
+                    {
+                        if (!argument.IsSuppressed)
+                        {
+                            if (!ReportNullableAssignmentIfNecessary(argument, parameterType, resultType,
+                                useLegacyWarnings: false, AssignmentKind.Argument, target: parameter))
+                            {
+                                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                                if (!_conversions.HasIdentityOrImplicitReferenceConversion(resultType.TypeSymbol, parameterType.TypeSymbol, ref useSiteDiagnostics))
+                                {
+                                    ReportNullabilityMismatchInArgument(argument, resultType.TypeSymbol, parameter, parameterType.TypeSymbol, forOutput: false);
+                                }
+                            }
+                        }
+                    }
+                    goto case RefKind.Out;
                 case RefKind.Out:
                     {
                         if (argument is BoundLocal local && local.DeclarationKind == BoundLocalDeclarationKind.WithInferredType)
                         {
                             _variableTypes[local.LocalSymbol] = parameterType;
-                            resultType = parameterType;
                         }
 
-                        TypeSymbolWithAnnotations adjustedParameterType = adjustNullableAnnotationForNullabilityCheck(parameterType, resultType);
+                        var lValueType = result.LValueType;
+                        var parameterValue = new BoundParameter(argument.Syntax, parameter);
```

Comment:
>parameterValue = new BoundParameter(argument.Syntax, parameter) [](start = 28, length = 63)

It's not clear why we need a `BoundParameter` here rather than using `argument`. Please add a comment.